File: .github/workflows/drivers/Handoff.java
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import packbin.Access;
import packbin.BinaryPacker;
import packbin.Packbin;
import packbin.Scheme;

public final class Handoff {
    public static void main(String[] args) {
        if (args.length < 1) {
            System.exit(2);
        }
        String cmd = args[0];
        switch (cmd) {
            case "pack-user" -> System.out.println(hex(BinaryPacker.pack(userScheme(), userValues())));
            case "pack-nested" -> System.out.println(hex(BinaryPacker.pack(nestedScheme(), nestedValues())));
            case "unpack-user" -> System.exit(userOk(requireHex(args)) ? 0 : 1);
            case "unpack-nested" -> System.exit(nestedOk(requireHex(args)) ? 0 : 1);
            default -> System.exit(2);
        }
    }

    private static String requireHex(String[] args) {
        if (args.length < 2) {
            System.exit(2);
        }
        return args[1];
    }

    @SuppressWarnings({"unchecked", "rawtypes"})
    private static Scheme<Map> userScheme() {
        return new Scheme<>(
                1,
                (Class) Map.class,
                Packbin.utf8(0, Access.get("username"), Access.set("username")),
                Packbin.list(Access.get("roles"), Access.set("roles"),
                        Packbin.utf8(0, Access.identity(), Access.ignore())),
                Packbin.dict(Access.get("access"), Access.set("access"),
                        Packbin.list(Access.identity(), Access.ignore(),
                                Packbin.utf8(0, Access.identity(), Access.ignore()))));
    }

    private static Map<String, Object> userValues() {
        Map<String, Object> access = new LinkedHashMap<>();
        access.put("channel", List.of("read"));
        access.put("map", List.of("read", "gps_fix", "set", "edit"));
        access.put("store", List.of("read", "write"));
        Map<String, Object> values = new HashMap<>();
        values.put("username", "zxsanny");
        values.put("roles", List.of("user", "dispatcher"));
        values.put("access", access);
        return values;
    }

    @SuppressWarnings({"unchecked", "rawtypes"})
    private static Scheme<Map> nestedScheme() {
        return new Scheme<>(
                1,
                (Class) Map.class,
                Packbin.dict(Access.get("access"), Access.set("access"),
                        Packbin.list(Access.identity(), Access.ignore(),
                                Packbin.dict(Access.identity(), Access.ignore(),
                                        Packbin.utf8(0, Access.identity(), Access.ignore())))));
    }

    private static Map<String, Object> nestedValues() {
        Map<String, Object> access = new LinkedHashMap<>();
        access.put("map", List.of(Map.of("op", "gps_fix")));
        access.put("store", List.of(Map.of("op", "read"), Map.of("op", "write")));
        Map<String, Object> values = new HashMap<>();
        values.put("access", access);
        return values;
    }

    @SuppressWarnings({"unchecked", "rawtypes"})
    private static boolean userOk(String hex) {
        Map[] got = new Map[1];
        Object err = BinaryPacker.unpack(parse(hex), userScheme().on(row -> got[0] = row));
        if (err != null || got[0] == null || got[0].size() != 3) {
            return false;
        }
        if (!"zxsanny".equals(got[0].get("username"))) {
            return false;
        }
        List<?> roles = (List<?>) got[0].get("roles");
        if (roles == null || roles.size() != 2 || !"user".equals(roles.get(0)) || !"dispatcher".equals(roles.get(1))) {
            return false;
        }
        Map<?, ?> access = (Map<?, ?>) got[0].get("access");
        return access != null
                && access.size() == 3
                && List.of("read").equals(access.get("channel"))
                && List.of("read", "gps_fix", "set", "edit").equals(access.get("map"))
                && List.of("read", "write").equals(access.get("store"));
    }

    @SuppressWarnings({"unchecked", "rawtypes"})
    private static boolean nestedOk(String hex) {
        Map[] got = new Map[1];
        Object err = BinaryPacker.unpack(parse(hex), nestedScheme().on(row -> got[0] = row));
        if (err != null || got[0] == null) {
            return false;
        }
        Map<?, ?> access = (Map<?, ?>) got[0].get("access");
        if (access == null || access.size() != 2) {
            return false;
        }
        return oneOp(access.get("map"), "gps_fix")
                && twoOps(access.get("store"), "read", "write");
    }

    private static boolean oneOp(Object raw, String op) {
        if (!(raw instanceof List<?> rows) || rows.size() != 1) {
            return false;
        }
        if (!(rows.get(0) instanceof Map<?, ?> fields)) {
            return false;
        }
        return op.equals(fields.get("op")) && fields.size() == 1;
    }

    private static boolean twoOps(Object raw, String a, String b) {
        if (!(raw instanceof List<?> rows) || rows.size() != 2) {
            return false;
        }
        return oneOp(List.of(rows.get(0)), a) && oneOp(List.of(rows.get(1)), b);
    }

    private static String hex(byte[] raw) {
        StringBuilder out = new StringBuilder();
        for (byte b : raw) {
            out.append(String.format("%02x", b));
        }
        return out.toString();
    }

    private static byte[] parse(String hex) {
        byte[] out = new byte[hex.length() / 2];
        for (int i = 0; i < out.length; i++) {
            out[i] = (byte) Integer.parseInt(hex.substring(i * 2, i * 2 + 2), 16);
        }
        return out;
    }
}

File: .github/workflows/drivers/handoff.ts
import { BinaryPacker, dict, list, scheme, utf8 } from "../../../typescript/src/index.ts";

type UserRow = {
  username: string;
  roles: string[];
  access: Record<string, string[]>;
};

type NestedRow = {
  access: Record<string, Record<string, string>[]>;
};

const userPacket = scheme<UserRow>(
  1,
  utf8(0, (r) => r.username),
  list(
    (r) => r.roles,
    utf8(0, (s) => s),
  ),
  dict(
    (r) => r.access,
    list(
      (a) => a,
      utf8(0, (s) => s),
    ),
  ),
);

const nestedPacket = scheme<NestedRow>(
  1,
  dict(
    (r) => r.access,
    list(
      (rows) => rows,
      dict(
        (row) => row,
        utf8(0, (s) => s),
      ),
    ),
  ),
);

const userValues: UserRow = {
  username: "zxsanny",
  roles: ["user", "dispatcher"],
  access: {
    channel: ["read"],
    map: ["read", "gps_fix", "set", "edit"],
    store: ["read", "write"],
  },
};

const nestedValues: NestedRow = {
  access: {
    map: [{ op: "gps_fix" }],
    store: [{ op: "read" }, { op: "write" }],
  },
};

function deepEqual(a: unknown, b: unknown): boolean {
  if (a === b) return true;
  if (Array.isArray(a) && Array.isArray(b)) {
    if (a.length !== b.length) return false;
    for (let i = 0; i < a.length; i++) {
      if (!deepEqual(a[i], b[i])) return false;
    }
    return true;
  }
  if (
    a !== null &&
    b !== null &&
    typeof a === "object" &&
    typeof b === "object" &&
    !Array.isArray(a) &&
    !Array.isArray(b)
  ) {
    const ao = a as Record<string, unknown>;
    const bo = b as Record<string, unknown>;
    const ak = Object.keys(ao).sort();
    const bk = Object.keys(bo).sort();
    if (ak.length !== bk.length) return false;
    for (let i = 0; i < ak.length; i++) {
      if (ak[i] !== bk[i]) return false;
      if (!deepEqual(ao[ak[i]!], bo[bk[i]!])) return false;
    }
    return true;
  }
  return false;
}

const cmd = process.argv[2];

if (cmd === "pack-user") {
  const bytes = BinaryPacker.pack(userPacket, userValues);
  process.stdout.write(Buffer.from(bytes).toString("hex") + "\n");
  process.exit(0);
}

if (cmd === "pack-nested") {
  const bytes = BinaryPacker.pack(nestedPacket, nestedValues);
  process.stdout.write(Buffer.from(bytes).toString("hex") + "\n");
  process.exit(0);
}

if (cmd === "unpack-user") {
  const hex = process.argv[3] ?? "";
  const bytes = Buffer.from(hex, "hex");
  let row: UserRow | undefined;
  const result = BinaryPacker.unpack(bytes, userPacket.on((value) => {
    row = value as UserRow;
  }));
  if (result.ok && row !== undefined && deepEqual(row, userValues)) process.exit(0);
  process.exit(1);
}

if (cmd === "unpack-nested") {
  const hex = process.argv[3] ?? "";
  const bytes = Buffer.from(hex, "hex");
  let row: NestedRow | undefined;
  const result = BinaryPacker.unpack(bytes, nestedPacket.on((value) => {
    row = value as NestedRow;
  }));
  if (result.ok && row !== undefined && deepEqual(row, nestedValues)) process.exit(0);
  process.exit(1);
}

process.exit(2);
